// upbit-js/dist/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Constant {
    constructor() {
        // quotation API
        this.UPBIT_URL_V1 = 'api.upbit.com/v1';
        this.UPBIT_URL_MARKET_ALL = 'market/all';
        this.UPBIT_URL_CANDLES_MINUTE = 'candles/minutes';
        this.UPBIT_URL_CANDLES_DAYS = 'candles/days';
        this.UPBIT_URL_CANDLES_WEEKS = 'candles/weeks';
        this.UPBIT_URL_CANDLES_MONTHS = 'candles/months';
        this.UPBIT_URL_TRADES_TICKS = 'trades/ticks';
        this.UPBIT_URL_TICKER = 'ticker';
        this.UPBIT_URL_ORDERBOOK = 'orderbook';
        // exchange API
        this.UPBIT_URL_ACCOUNTS = 'accounts';
        this.UPBIT_URL_ORDERS_CHANCE = 'orders/chance';
        this.UPBIT_URL_ORDER = 'order';
        this.UPBIT_URL_ORDERS = 'orders';
        this.UPBIT_URL_WITHRAWS = 'withdraws';
        this.UPBIT_URL_WITHRAW = 'withdraw';
        this.UPBIT_URL_WITHRAWS_CHANCE = 'withdraws/chance';
        this.UPBIT_URL_WITHRAWS_COIN = 'withdraws/coin';
        this.UPBIT_URL_WITHRAWS_KRW = 'withdraws/krw';
        this.UPBIT_URL_DEPOSITS = 'deposits';
        this.UPBIT_URL_DEPOSIT = 'deposit';
        this.UPBIT_URL_DEPOSITS_GENERATE_COIN_ADDRESS = 'deposits/generate_coin_address';
        this.UPBIT_URL_DEPOSITS_COIN_ADDRESSES = 'deposits/coin_addresses';
        this.UPBIT_URL_DEPOSITS_COIN_ADDRESS = 'deposits/coin_address';
    }
    static GetConstant() {
        if (!this.constant)
            this.constant = new Constant();
        return this.constant;
    }
}
Constant.constant = undefined;
exports.Constant = Constant;
exports.constant = Constant.GetConstant();
